### PR TITLE
CBottle Video Model

### PR DESCRIPTION
**Pull request by NickGeneva**
# Earth2Studio Pull Request

## Description

Adds CBottle Video as a prognostic model that allows users to unroll cbottle forecasts.

Blocked by: https://github.com/NVIDIA/earth2studio/pull/440

Closes: https://github.com/NVIDIA/earth2studio/issues/433

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/earth2studio/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/earth2studio/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/earth2studio/issues) is linked to this pull request.

### CODE REVIEW

**Comment by NickGeneva** [comment]
Here is how I validated the model against the vanilla training loader and sampler

```python
import numpy as np
import torch
from collections import OrderedDict

import os
os.environ["AMIP_MID_MONTH_SST"] = "/localhome/local-ngeneva/earth2studio/cbottle_v1.2/amip_midmonth_sst.nc"

from cbottle.inference import CBottle3d
from cbottle.visualizations import visualize
from cbottle.datasets import dataset_3d
from cbottle.training.video.frame_masker import FrameMasker
import matplotlib.pyplot as plt

import earth2grid
from datetime import datetime
from earth2studio.models.auto import Package
from earth2studio.models.px import CBottleVideo

# Input parameters
start_time = datetime(1990,1,1)
lat_grid = np.linspace(90, -90, 721)
lon_grid = np.linspace(0, 360, 1440, endpoint=False)

# Sample using built in methods from cBottle
torch.manual_seed(0)
torch.cuda.manual_seed(0)
dataset = dataset_3d.get_dataset(
    split="",
    dataset="amip",
    sst_input=True,
    infinite=False,
    shuffle=False,
    time_step=6,
    time_length=12,
    frame_masker=FrameMasker(keep_frames=[]),
    variable_config=dataset_3d.VARIABLE_CONFIGS["default"],
    map_style=True,
)
# Find index closest to start_time
time_diffs = [abs(t - start_time) for t in dataset.times]
closest_idx = time_diffs.index(min(time_diffs))
print(f"Closest time index: {closest_idx}")
# loader = torch.utils.data.DataLoader(dataset, batch_size=1)
batch0 = dataset.__getitem__(closest_idx)

for key, value in batch0.items():
    if isinstance(value, torch.Tensor):
        batch0[key] = value.unsqueeze(0).cuda()

model_path = os.path.join("/localhome/local-ngeneva/earth2studio/cbottle_v1.2", "cBottle-video.zip")
model = CBottle3d.from_pretrained([model_path], sigma_min=0.02, sigma_max=1000.0)
output, coords_0 = model.sample(batch0, seed=0)

# Convert into lat/lon
nlat, nlon = 721, 1440
latlon_grid = earth2grid.latlon.equiangular_lat_lon_grid(
    nlat, nlon, includes_south_pole=True
)
regridder = earth2grid.get_regridder(coords_0.grid, latlon_grid).to("cuda")
output_org = regridder(output).squeeze(2).transpose(1,2)

print("================================")

torch.manual_seed(0)
torch.cuda.manual_seed(0)
model2 = CBottleVideo.load_model(Package("file:///localhome/local-ngeneva/earth2studio/cbottle_v1.2"), seed=0).to("cuda")
coords = OrderedDict({
    "time": np.array([start_time], dtype=np.datetime64),
    "lead_time": np.array([np.timedelta64(0)]),
    "variable": np.array(model2.VARIABLES),
    "lat": lat_grid,
    "lon": lon_grid
})
inputs = torch.full((1,1,45,721,1440), float('nan')).cuda()
# Execute prognostic
outputs = []
model_itr = model2.create_iterator(inputs, coords)
for step, (x, coords) in enumerate(model_itr): 
    outputs.append(x.cpu())
    if step == 11:
        break
out = torch.cat(outputs, dim=1)

assert torch.allclose(output_org.cpu(), out.cpu())
```

**Comment by NickGeneva** [comment]
For the HEALPIX support this is the respective test (plus different seed):

```python
import numpy as np
import torch
from collections import OrderedDict

import os
os.environ["AMIP_MID_MONTH_SST"] = "/localhome/local-ngeneva/earth2studio/cbottle_v1.2/amip_midmonth_sst.nc"

from cbottle.inference import CBottle3d
from cbottle.visualizations import visualize
from cbottle.datasets import dataset_3d
from cbottle.training.video.frame_masker import FrameMasker
import matplotlib.pyplot as plt

import earth2grid
from datetime import datetime
from earth2studio.models.auto import Package
from earth2studio.models.px import CBottleVideo

# Input parameters
start_time = datetime(1990,1,1)
lat_grid = np.linspace(90, -90, 721)
lon_grid = np.linspace(0, 360, 1440, endpoint=False)

# Sample using built in methods from cBottle
torch.manual_seed(0)
torch.cuda.manual_seed(0)
dataset = dataset_3d.get_dataset(
    split="",
    dataset="amip",
    sst_input=True,
    infinite=False,
    shuffle=False,
    time_step=6,
    time_length=12,
    frame_masker=FrameMasker(keep_frames=[]),
    variable_config=dataset_3d.VARIABLE_CONFIGS["default"],
    map_style=True,
)
# Find index closest to start_time
time_diffs = [abs(t - start_time) for t in dataset.times]
closest_idx = time_diffs.index(min(time_diffs))
print(f"Closest time index: {closest_idx}")
# loader = torch.utils.data.DataLoader(dataset, batch_size=1)
batch0 = dataset.__getitem__(closest_idx)

for key, value in batch0.items():
    if isinstance(value, torch.Tensor):
        batch0[key] = value.unsqueeze(0).cuda()

model_path = os.path.join("/localhome/local-ngeneva/earth2studio/cbottle_v1.2", "cBottle-video.zip")
model = CBottle3d.from_pretrained([model_path], sigma_min=0.02, sigma_max=1000.0)
output_org, coords_0 = model.sample(batch0, seed=1)
output_org = output_org.transpose(1,2)

print("================================")

torch.manual_seed(0)
torch.cuda.manual_seed(0)
model2 = CBottleVideo.load_model(Package("file:///localhome/local-ngeneva/earth2studio/cbottle_v1.2"), lat_lon=False, seed=1).to("cuda")
coords = OrderedDict({
    "time": np.array([start_time], dtype=np.datetime64),
    "lead_time": np.array([np.timedelta64(0)]),
    "variable": np.array(model2.VARIABLES),
    "hpx": np.arange(49152)
})
inputs = torch.full((1,1,45,49152), float('nan')).cuda()
# Execute prognostic
outputs = []
model_itr = model2.create_iterator(inputs, coords)
for step, (x, coords) in enumerate(model_itr): 
    print(x.shape)
    outputs.append(x.cpu())
    if step == 11:
        break
out = torch.cat(outputs, dim=1)

assert torch.allclose(output_org.cpu(), out.cpu())
```

**Comment by NickGeneva** [comment]
Sample:

```python
import torch
import numpy as np

from collections import OrderedDict
from datetime import datetime
from earth2studio.models.auto import Package
from earth2studio.models.px import CBottleVideo

start_time = datetime(1990,1,1)

model = CBottleVideo.load_model(CBottleVideo.load_default_package(), seed=1).to("cuda")
coords = OrderedDict({
    "time": np.array([start_time], dtype=np.datetime64),
    "lead_time": np.array([np.timedelta64(0, 'h')]),
    "variable": np.array(model.VARIABLES),
    "lat": np.linspace(90, -90, 721),
    "lon": np.linspace(0, 360, 1440, endpoint=False),
})
inputs = torch.full((1,1,45,721,1440), float('nan')).cuda()
# Execute prognostic
outputs = []
model_itr = model.create_iterator(inputs, coords)
for step, (x, coords) in enumerate(model_itr): 
    print(x.shape)
    outputs.append(x.cpu())
    if step == 11:
        break
out = torch.cat(outputs, dim=1).cpu().numpy()


# Post processing!

import matplotlib.pyplot as plt
vidx = -13
n_frames = out.shape[1]

fig, axes = plt.subplots(3, 4, figsize=(16, 12))
axes = axes.flatten()

for i in range(n_frames):
    axes[i].imshow(out[0,i,vidx,:])
    axes[i].set_title(f'Frame {i}')
    axes[i].axis('off')

plt.tight_layout()
plt.savefig("cbottle.jpg")
```
cbottle

File: earth2studio/models/px/cbottle_video.py [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+            [(t.hour * 3600) + (t.minute * 60) + t.second for t in times0]
+        ).reshape(times.shape)
+        day_of_year = np.array(
+            [(t - datetime(t.year, 1, 1)).total_seconds() / (86400.0) for t in times0]
+        ).reshape(times.shape)
+        second_of_day = torch.tensor(second_of_day.astype(np.float32), device=device)
+        day_of_year = torch.tensor(day_of_year.astype(np.float32), device=device)
+
+        # Target tensor, not needed for video model since no infill
+        # target = torch.zeros(
```
Comment (loliverhennigh) [inline earth2studio/models/px/cbottle_video.py:364]: Old code or did you just want to leave this for reference?

File: earth2studio/models/px/cbottle_video.py [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+
+    Note
+    ----
+    This wrapper allows users to provide an input condition for the first frame of the
+    model. If this tensor is all NaNs no variable conditioning will be used running the
+    network outside of time-stamp and respective SST.
+
+    Warning
+    -------
+    Default model package has SST data from January 1940 to December 2022, expanded SST
```
Comment (loliverhennigh) [inline earth2studio/models/px/cbottle_video.py:72]: Might be good to add a check for this in the model

File: earth2studio/models/px/cbottle_video.py [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+        Core Pytorch model
+    sst_ds : xr.Dataset
+        Sea surface temperature xarray dataset
+    lat_lon : bool, optional
+        Lat/lon toggle, if true data source will return output on a 0.25 deg lat/lon
+        grid. If false, the native nested HealPix grid will be returned, by default True
+    sampler_steps : int, optional
+        Number of diffusion steps, by default 18
+    sigma_max : float, optional
+        Noise amplitude used to generate latent variables, by default 80
```
Comment (loliverhennigh) [inline earth2studio/models/px/cbottle_video.py]: This mismatches default

File: earth2studio/models/px/cbottle_video.py (hunk 1 above)
Comment (NickGeneva) [reply earth2studio/models/px/cbottle_video.py:364]: Left for reference so I know what is typically there in the cbottle repo

File: earth2studio/models/px/cbottle_video.py (hunk 2 above)
Comment (NickGeneva) [reply earth2studio/models/px/cbottle_video.py:72]: Added a warning so people could theoretically add different ssts

File: earth2studio/models/px/cbottle_video.py (hunk 3 above)
Comment (NickGeneva) [reply earth2studio/models/px/cbottle_video.py]: Good catch, fixed!

**Review by loliverhennigh — APPROVED** [review]
